upup/pkg/fi/cloudup/openstacktasks/sshkey.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openstacktasks
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/keypairs"
+
+	"k8s.io/kops/pkg/pki"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/cloudup/openstack"
+)
+
+//go:generate fitask -type=SSHKey
+type SSHKey struct {
+	Name      *string
+	Lifecycle *fi.Lifecycle
+
+	PublicKey *fi.ResourceHolder
+
+	KeyFingerprint *string
+}
+
+var _ fi.CompareWithID = &SSHKey{}
+
+func (e *SSHKey) CompareWithID() *string {
+	return e.Name
+}
+
+func (e *SSHKey) Find(c *fi.Context) (*SSHKey, error) {
+	cloud := c.Cloud.(openstack.OpenstackCloud)
+	rs, err := cloud.ListKeypair(openstackKeyPairName(fi.StringValue(e.Name)))
+	if err != nil {
+		return nil, err
+	}
+	if rs == nil {
+		return nil, nil
+	}
+	actual := &SSHKey{
```

Comment:
Sorry, I am not quite understanding these codes. For me, I will implement it like this.
if rs.PublicKey != e.PublicKey {
    return nil, fmt.Errorf("the public key is not match")
}
if e.KeyFingerprint != nil && rs.KeyFingerprint != e.KeyFingerprint {
    return nil, fmt.Errof("the fingerprint is not match")
}
actual := &SSHKey{
    Name:           e.Name,
    KeyFingerprint: fi.String(rs.Fingerprint),
    PublicKey: fi.String(rs.PublicKey)
}

Reply:
this is also copypasted from aws.

Reply:
Ah - so Find should return the cloud object (SSH key) as it exists.   So for example if we're creating a VPC on AWS, we return the "matching" VPC, but the values might not be correct, and `Render` would either create a VPC (if no VPC was found), or it would change the VPC to match the expected value.

The code is particularly odd on AWS because AWS keypairs can't be changed once created, and we also _only_ get the fingerprint.  It looks like OpenStack returns the public key, which means we could skip the weird "if the fingerprint matches assume the public key matches" logic from AWS.

So I think this code is right, but we might be able to do better on OpenStack